docs/spec/auth/vesting.md

```diff
@@ -74,28 +58,46 @@ type BaseVestingAccount struct {
 
     OriginalVesting  Coins // coins in account upon initialization
     DelegatedFree    Coins // coins that are vested and delegated
-    EndTime          Time // when the coins become unlocked
+    DelegatedVesting Coins // coins that vesting and delegated
+
+    EndTime  Time // when the coins become unlocked
 }
 
 // ContinuousVestingAccount implements the VestingAccount interface. It
 // continuously vests by unlocking coins linearly with respect to time.
 type ContinuousVestingAccount struct {
-    BaseAccount
     BaseVestingAccount
 
-    DelegatedVesting Coins // coins that vesting and delegated
-    StartTime        Time // when the coins start to vest
+    StartTime  Time // when the coins start to vest
 }
 
 // DelayedVestingAccount implements the VestingAccount interface. It vests all
 // coins after a specific time, but non prior. In other words, it keeps them
 // locked until a specified time.
 type DelayedVestingAccount struct {
-    BaseAccount
     BaseVestingAccount
 }
 ```
 
+In order to facilitate less ad-hoc type checking and assertions and to support
```

Comment:
It would be better for x/bank to use runtime type checking to check for a vesting-account interface, than to require the auth Account interface to implement SpendableCoins/TrackDelegation/TrackUndelegation.  These methods should be declared in VestingAccount.  x/auth should know nothing of vesting, it should all be in x/bank.

If this isn't possible, please let me know why not.  If it is possible, we should probably do it, and not require x/auth accounts to know about vesting.

Reply:
I think it is possible - previously this was done (in some form) - but it made the bank code far more difficult to read and reason through, so we decided to abstract away this functionality in the account interface, which made it much easier to reason through.

Would moving `VestingAccount` to `x/bank` (but retaining the interface abstraction) help?

Reply:
@cwgoes and I wanted to avoid type checking (as that was the way it was initially) as that pollutes the code and makes it less flexible.

>  x/auth should know nothing of vesting, it should all be in x/bank.

It would technically, as it would have to type check.

> If this isn't possible, please let me know why not. If it is possible, we should probably do it, and not require x/auth accounts to know about vesting.

It is possible and that was the way it was initially intended, but there was a bit of dup code and was not as clean. I do not want to keep going back-n-forth on this and just simply decide as this PR has been open for a long time now.

Reply:
I vaguely recall us having a phone convo about this 🤔 Wouldn't it also be weird to have a `VestingAccount` in `x/bank` as bank should not really define account types?